src/constraint_system/standard/composer.rs

```diff
@@ -266,11 +266,17 @@ impl Composer for StandardComposer {
         //1. Compute witness Polynomials
         //
         // Convert Variables to Scalars
-        // XXX: Maybe there's no need to allocate `to_scalars` returning &[Scalar].
-        let w_l_scalar = self.to_scalars(&self.w_l);
-        let w_r_scalar = self.to_scalars(&self.w_r);
-        let w_o_scalar = self.to_scalars(&self.w_o);
-        let w_4_scalar = self.to_scalars(&self.w_4);
+        let mut w_l_scalar = self.to_scalars(&self.w_l);
+        let mut w_r_scalar = self.to_scalars(&self.w_r);
+        let mut w_o_scalar = self.to_scalars(&self.w_o);
+        let mut w_4_scalar = self.to_scalars(&self.w_4);
+
+        // Pad witness
+        let pad = vec![Scalar::zero(); domain.size() - w_l_scalar.len()];
+        w_l_scalar.extend(pad.iter());
```

Comment:
It will be easier to concat slices than extend vectors since we save re-allocations and also will be more performant and less memmory consuming.

Reply:
I think the witnesses should be the same size everywhere, so we would need to do it here. I think the compiler is smart enough to optimise the Iterator when using extend, or are there indications that this is a problem with large circuits?

Reply:
TBH idk if it will be optimized. 

But the changes in 62af0c0 produce cleaner code and also force the compiler to store these refs instead of letting it decide.

We can benchmark if you want, but dunno if it's even worth the time.